themes/default/content/blog/pulumi-is-imperative-declarative-imperative/index.md

```diff
@@ -0,0 +1,221 @@
+     Type                    Name           Status
+ +   pulumi:pulumi:Stack     blog_code-dev  created
+ +   ├─ aws:s3:Bucket        bucket         created
+ +   ├─ aws:s3:BucketObject  object-0       created
+ +   ├─ aws:s3:BucketObject  object-1       created
+ +   ├─ aws:s3:BucketObject  object-3       created
+ +   ├─ aws:s3:BucketObject  object-2       created
+ +   ├─ aws:s3:BucketObject  object-6       created
+ +   ├─ aws:s3:BucketObject  object-4       created
+ +   ├─ aws:s3:BucketObject  object-5       created
+ +   ├─ aws:s3:BucketObject  object-7       created
+ +   ├─ aws:s3:BucketObject  object-9       created
+ +   └─ aws:s3:BucketObject  object-8       created
+
+Resources:
+    + 12 created
+
+Duration: 13s
+```
+
+While this is definitely an imperative program, there is one important thing to understand: instantiating an
+`s3.Bucket`, `s3.BucketObject` or any other Pulumi resource should not be interpreted as an imperative creation
+of the resource in the language host. Behind the scenes, any resource instantiation in the language host triggers a
+`Register Resource` request to the Pulumi engine. All these `Register Resource` requests together form the
+resource model you as an infrastructure developer want to get in the end.
+
+Running your program always sends the full resource model to the Pulumi Engine regardless of what state your
+current infrastructure is in.
+
+Our previous example was shown in Python, but recently we also delivered [support for YAML](https://www.pulumi.com/docs/intro/languages/yaml/).
+The creation of our S3 bucket could be converted to this snippet:
+
+```yaml
+resources:
+  bucket:
+    type: aws:s3:Bucket
+```
+
+The separation of language support from the engine allows us to offer both imperative & declarative solutions.
+But remember that none of the supported languages do any of the provisioning directly in the language host.
+
+If the resource provisioning is not taking place in the language host, where is the magic happening then?
+
+### CLI and Engine
+
+In the previous step, you found out that the language host sends requests to the engine to fullfil your to-be
+infrastructure.
+
+It is now that the Pulumi engine gets to work. The engine combines the intended model of the infrastructure
+received from the language host, the current state recorded in the state backend and the actual resource state
+to compute which actions need to be executed to bring the actual state in line with the intended model.
+
+Our little example contains dependencies:
+every `s3.BucketObject` uses the `bucket.id` as way to define in which bucket these objects should be stored.
+The property `id` from the `s3.Bucket` is an `Output`. Outputs are Pulumi's way of tracking which property of one resource
+is required by another, hereby creating a dependency between the resources. The engine uses all these outputs
+passed from one resource to another as the vertices in a directed acyclic graph (DAG). The engine determines
```

Comment:
May I recommend linking to something that explains DAGs? A number of folks won't know what that means.